packages/react-dom/src/client/ReactDOMComponent.js

```diff
@@ -248,6 +208,53 @@ if (__DEV__) {
   };
 }
 
+// HTML parsing normalizes CR and CRLF to LF.
+// It also can turn \u0000 into \uFFFD inside attributes.
+// https://www.w3.org/TR/html5/single-page.html#preprocessing-the-input-stream
+// If we have a mismatch, it might be caused by that.
+// We will still patch up in this case but not fire the warning.
+const NORMALIZE_NEWLINES_REGEX = /\r\n?/g;
+const NORMALIZE_NULL_AND_REPLACEMENT_REGEX = /\u0000|\uFFFD/g;
+
+function normalizeMarkupForTextOrAttribute(markup: mixed): string {
+  if (__DEV__) {
+    checkHtmlStringCoercion(markup);
+  }
+  const markupString = typeof markup === 'string' ? markup : '' + (markup: any);
+  return markupString
+    .replace(NORMALIZE_NEWLINES_REGEX, '\n')
+    .replace(NORMALIZE_NULL_AND_REPLACEMENT_REGEX, '');
+}
+
+export function checkForUnmatchedText(
+  serverText: string,
+  clientText: string | number,
+  isConcurrentMode: boolean,
+) {
+  const normalizedClientText = normalizeMarkupForTextOrAttribute(clientText);
```

Comment:
Since this is not prod behavior it’s a bit more perf sensitive. We can first check if the original string is equal first, which it in most cases will be and if so bail early. Only if that is not true should we normalize.

Reply:
Oh nvm. We only get here if they’re not already equal.